File: src/main/java/dev/boooiil/historia/core/items/component/RunnableComponent.java
package dev.boooiil.historia.core.items.component;

import dev.boooiil.historia.core.items.ItemComponent;
import dev.boooiil.historia.core.items.data.RunnableData;
import dev.boooiil.historia.core.util.JSONUtils;
import org.bukkit.configuration.ConfigurationSection;

public class RunnableComponent implements ItemComponent {

    private final int ticks;
    private final String command;
    private final String permission;

    public RunnableComponent(int ticks, String command, String permission) {
        this.ticks = ticks;
        this.command = command;
        this.permission = permission;
    }

    public static RunnableComponent fromConfig(ConfigurationSection section) {

        int ticks = section.getInt("ticks");
        String command = section.getString("command");
        String permission = section.getString("permission");

        return new RunnableComponent(ticks, command, permission);

    }

    @Override
    public RunnableData data() {
        return new RunnableData(this.ticks, this.command, this.permission);
    }

    @Override
    public RunnableData data(float qualityModifier) {
        return data();
    }

    @Override
    public String getKey() {
        return "runnable";
    }

    @Override
    public String toString() {

        String sb = "RunnableComponent" +
                "{" +
                JSONUtils.fromValue("ticks", ticks) + ", " +
                JSONUtils.fromValue("command", command) + ", " +
                JSONUtils.fromValue("permission", permission) +
                "}";

        return sb;

    }

    @Override
    public String toJSON() {

        String sb = "{" +
                JSONUtils.fromValue("ticks", ticks) + ", " +
                JSONUtils.fromValue("command", command) + ", " +
                JSONUtils.fromValue("permission", permission) +
                "}";

        return sb;

    }

}

File: src/main/java/dev/boooiil/historia/core/items/component/RunnableComponent.kt
package dev.boooiil.historia.core.items.component

import dev.boooiil.historia.core.items.ItemComponent
import dev.boooiil.historia.core.items.data.RunnableData
import dev.boooiil.historia.core.util.JSONUtils
import org.bukkit.configuration.ConfigurationSection

class RunnableComponent(
    private val ticks: Int,
    private val command: String,
    private val permission: String
) : ItemComponent {

    override val key = "runnable"

    override fun data(qualityModifier: Double?): RunnableData {
        return RunnableData(this.ticks, this.command, this.permission)
    }

    override fun toString(): String {
        val sb = "RunnableComponent" +
                "{" +
                JSONUtils.fromValue("ticks", ticks) + ", " +
                JSONUtils.fromValue("command", command) + ", " +
                JSONUtils.fromValue("permission", permission) +
                "}"

        return sb
    }

    override fun toJSON(): String {
        val sb = "{" +
                JSONUtils.fromValue("ticks", ticks) + ", " +
                JSONUtils.fromValue("command", command) + ", " +
                JSONUtils.fromValue("permission", permission) +
                "}"

        return sb
    }

    companion object {
        fun fromConfig(section: ConfigurationSection): RunnableComponent {
            val ticks = section.getInt("ticks")
            val command = section.getString("command")
            val permission = section.getString("permission")

            return RunnableComponent(ticks, command!!, permission!!)
        }
    }
}
